proposals/interface-types/working-notes/Instructions.md

```diff
@@ -0,0 +1,252 @@
+# Interface instructions
+
+This proposal defines new instructions that operate over interface value types. These instructions have the same form as core instructions; they push and pop values on an execution stack and support single pass validation.
+
+The new interface instructions are summarized below.
+
+```
+#interface-instr ::=
+  | bool.lift_iMM
+  | bool.lower_iMM
+
+  | sNN.lift_iMM
+  | sNN.lower_iMM
+
+  | uNN.lift_iMM
+  | uNN.lower_iMM
+
+  | string.lift_memory ..
+  | string.lower_memory ..
+
+  | record.lift ..
+  | record.lower ..
+
+  | array.lift_memory ..
+  | array.lower_memory ..
+
+  | invoke ..
+
+where:
+  NN is one of [8, 16, 32, 64]
+  MM is one of [32, 64]
+```
+
+In addition the following instructions are included from the core instructions:
+
+```
+#interface-instr ::=
+  | ..
+  | <numeric instructions>
+  | <memory instructions>
+  | <table instructions>
+  | drop
+  | unreachable
+  | nop
+```
+
+Refer to the definition of instruction classes [here](https://webassembly.github.io/spec/core/syntax/instructions.html).
+
+## Laziness
+
+Interface values generated by interface instructions are lazy. The execution of an instruction generating an interface value happens immediately before the instruction that consumes the value.
+
+## `bool.lift_iMM`
+
+This instruction has type `[iMM] -> [bool]`.
+
+The output is `true` iff the input `iMM` has any bit set in the integer.
+
+## `bool.lower_iMM`
+
+This instruction has type `[bool] -> iMM`.
+
+The output is `0b1` zero extended to the desired bitwidth iff the input is true.
+
+## `sNN.lift_iMM`
+
+This instruction has type `iMM` -> `sNN`.
+
+The output is the value of `iMM` as a two's complement signed integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `sNN.lower_iMM`
+
+This instruction has type `[sNN] -> [iMM]`
+
+The output is the value of `sNN` as a two's complement signed integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `uNN.lift_iMM`
+
+This instruction has type `iMM` -> `uNN`.
+
+The output is the value of `iMM` as an unsigned integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `uNN.lower_iMM`
+
+This instruction has type `[uNN] -> [iMM]`
+
+The output is the value of `uNN` as an unsigned integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `string.lift_memory`
+
+The full syntax of this instruction is:
+```
+string.lift_memory #memidx #encoding
+```
+
+The type of this instruction is:
+```
+[$base: i32, $len: i32] -> [string]
+```
+
+The output is a sequence of unicode code points taken by interpreting `store.memory[$memidx].bytes[$base .. $base + $len]` with `#encoding`.
+
+This instructions traps if the range `$base .. $base + $len` is out-of-bounds, or the byte sequence is not valid under `#encoding`.
+
```

Comment:
rather than trapping, allow a break in case of errors.
This will need more thought; there are use cases where just parsing the byte code is considered too expensive.